sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs

```diff
@@ -32,29 +32,1185 @@ public partial class SharedAccessSignatureCredential
 }
 namespace Azure.Messaging.EventGrid.Models
 {
-    public partial class CloudEvent
+    public partial class CloudEvent : System.Collections.Generic.ICollection<System.Collections.Generic.KeyValuePair<string, object>>, System.Collections.Generic.IDictionary<string, object>, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object>>, System.Collections.IEnumerable
```

Comment:
Hmm, what is this all about?

Reply:
Uh I'm not too sure what's going on there, but I know CloudEvent implements IDictionary<string, object> due to the AdditionalProperties property

Reply:
You are right, but I would have expected us to have a dictionary property instead.
@ellismg is this what you are intending with `additionalProperties` - making the `CloudEvent` itself act as a dictionary?

Reply:
> You are right, but I would have expected us to have a dictionary property instead.
@ellismg is this what you are intending with additionalProperties - making the CloudEvent itself act as a dictionary?

So, to take a step back and provide context here let me explain the problem we're trying to solve (my apologies if this is review for the both of you).

The CloudEvents spec talks about something called ["Extension Context Attributes"](https://github.com/cloudevents/spec/blob/v1.0/spec.md#extension-context-attributes) which allows the producer of a cloud event to add properties to an event.  This spec talks about the abstract design for the CloudEvent "object" but doesn't concern itself with how it is serliazied into JSON.

When the event is serialized into JSON, these key values pairs become properties on the event envelope (along with the known properties like `Id`, `Source`, etc.).  This `CloudEvent` type here represents the JSON encoding of a cloud event (which is described in a [different specification](https://github.com/cloudevents/spec/blob/v1.0/json-format.md), which is also why it has both a `Date` and `DataBase64` property (the general cloud event specification does not talk about `DataBase64`, rather that's something specific to how you can send binary data for an event when encoding it as JSON).

The use of additional properties in the swagger definition for CloudEvent was so that the generated model type would have a place for us to put these extension attributes and would ensure that the generated serialization code from Auto Rest would ensure that these properties would be included on the envelope itself (if we instead had said there was an `ExtensionAttributes` property on the model of type `Dictionary<string, object>`, it would mean that the JSON object that we serialized to would have some top level property called `extensionAttributes` which would be set to an object, but that's not what we want.

So, because of the above, I think what ever type we end up passing to the generated `sendCloudEvents` function will need to have this shape.

But I agree that this ends up being a very confusing model to present to end users.  In addition to the fact that CloudEvent is-a Dictionary (so there can be additional properties) having both `Data` and `DataBase64` properties is confusing, and probably not something we want to expose to end users, since it leaks an implementation detail about the JSON event format that we use to send the event.

In TypeScript, the "model" type I expose for a `CloudEvent` is actually not the same one generated by swagger.  Instead, it's a hand authored type that describes the shape of a cloud event as described by the general specification (so I don't have both `data` and `dataBase64` properties, just a `data` property).  The convivence layer ends up [converting](https://github.com/Azure/azure-sdk-for-js/blob/9f96bef22e46ea4296b2b6bc7493d0e4c0135bbf/sdk/eventgrid/eventgrid/src/eventGridClient.ts#L170-L211) from this public format into the format expected by the generated client (which we don't expose publicly from the TypeScript package).  Because of this translation layer, it means that I have full control over the shape of the public type.  Because of that, I was able to add an `extensionAttributes` dictionary to the [exported model type](https://github.com/Azure/azure-sdk-for-js/pull/10059/files#diff-c3aed8482f9780ccb5ee606d3cef4ea1R48) and then copy these properties onto the [generated model type](https://github.com/Azure/azure-sdk-for-js/pull/10059/files#diff-9ea0967503aea374f2f383a8b940951dR53).

In addition to being able to not expose the `Data`/`DataBase64` the other reason that pushed us towards this split was that when generating code for a model that has `additionalProperties`, the code generator adds `[property: string]: any;` to the generated type.  This basically says "in additional to all these other listed properties, you can have any other named properties on the object".  But exposing a model with that shape is dangerous for us because it means that TypeScript won't catch "typos" since there's no difference between `obj.prop` and `obj["prop"]`, which means if you made a typo in a known property name (but one that was optional) there'd be no way for TypeScript to detect that you had made a mistake.

So the long and short of it is, I think the code needs to look this way for the type we pass to the generated client.  I don't think that necessarily means that we need to expose this to end users however, we could hand author a `CloudEvent` type in the models package that has the shape we want, rename this to `CloudEventInternal` or something and just make it internal (I think that's supported by the tooling) and then have a conversion layer between the two.

Does this help clear things up?

Reply:
Yep, that is very helpful, thanks - this is the necessary shape for serialization, but we would likely want to expose a different shape to users. I believe we were planning on having a hand written CloudEvent anyway to handle the defaulting of Id/Time.